File: NXTBot/src/uk/co/shanksi/nxt/NxtBrick.java
package uk.co.shanksi.nxt;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import uk.co.shanksi.nxt.NxtCommand;
import uk.co.shanksi.nxt.SharedConstants;
import uk.co.shanksi.nxt.CommandType;

public class NxtBrick {
	
	private OutputStream out;
	private InputStream in;
	
	
	/**
	 * The constructor needs both io streams 
	 * to communicate with a NXT brick.
	 *
	 * @param out the stream to send commands to
	 * @param in the stream to read responses from
	 */
	public NxtBrick(OutputStream out, 
			InputStream in) {
		this.out = out;
		this.in = in;
//try{
      //  playTone('d', 'z');
  //  setOutputState((byte)0x02, (byte)0x55, true, false, (byte)0x20);
	
	//	} catch (IOException ex) {}
		}
	







	public void setOutputState(byte motor, 
			byte power, 
			boolean speedReg, 
			boolean motorSync, 
			byte runState) throws IOException {
		byte[] msg = { (byte)0x80, 0x04, motor, power, 
				0x07, 0x00, 0x00, runState, 0x00, 
				0x00, 0x00, 0x00 };
 
		sendMessage(msg);
	}


    /**
     * Sets the output state of a specific sensor port.
     */
    public synchronized void setOutputState(int portId, byte power, int mode,
                                               int regulationMode, int turnRatio,
                                               int runState, long tachoLimit)
    {
/*    if (debugLevel >= DEBUG_LEVEL_MEDIUM)
      DebugConsole.show("DEBUG: setOutputState("
        + portId + ", "
        + power + ", "
        + mode + ", "
        + regulationMode + ", "
        + turnRatio + ", "
        + runState + ", "
        + tachoLimit + ")");*/
        byte[] request = {
	        CommandType.DIRECT_COMMAND_NOREPLY, 
			NxtCommand.SET_OUTPUT_STATE, 
			(byte)portId,
            power, 
			(byte)mode, 
			(byte)regulationMode,
            (byte)turnRatio, 
			(byte)runState, 
			(byte)tachoLimit,
            (byte)(tachoLimit >>> 8), 
			(byte)(tachoLimit >>> 16),
			(byte)(tachoLimit >>> 24)
        };
		try {
            sendMessage(request);
		} 
		catch(IOException ex) {}
    }



    public void setInputMode(int portId, int sensorType, int sensorMode)
    {
/*    if (debugLevel >= DEBUG_LEVEL_MEDIUM)
      DebugConsole.show("DEBUG: setInputMode("
        + portId + ", "
        + sensorType + ", "
        + sensorMode + ")");*/
        byte[] request =
                {
                        CommandType.DIRECT_COMMAND_NOREPLY, NxtCommand.SET_INPUT_MODE, (byte)portId,
                        (byte)sensorType, (byte)sensorMode
                };
				try{
        sendMessage(request);} 
	catch(IOException ex) {}
    }

    /**
     * Reads the data returned as a reply from the brick.
     * @return the brick's reply
     */
    public synchronized byte[] readData()
    {
        byte[] reply = null;
        int length = -1;
        int lenMSB;
        int lenLSB;

        try
        {
            do
                lenLSB = in.read();
            while (lenLSB < 0);

            lenMSB = in.read(); // MSB of reply length
            length = (0xFF & lenLSB) | ((0xFF & lenMSB) << 8);

            reply = new byte[length];
            // Rest of packet
            in.read(reply);
        }
        catch (IOException ex)
        {
//      System.out.println("read error");
        }

      //  if (debugLevel == DEBUG_LEVEL_HIGH)
    //    {
 /*     DebugConsole.show("DEBUG: readData() returned:");
      for (int i = 0; i
        < reply.length; i++)
        DebugConsole.show("  " + reply[i]);
    *///}

        return reply;
    }


    public synchronized byte[] requestData(byte[] request)
{
try{	
	sendMessage(request);} 
catch(IOException ex) {}
	return readData();
}
/**
 * Reads the values from given a sensor port.
 * @param portId the Id of the sensor port (0,..3)
 */
  public synchronized InputValues getInputValues(int portId)
  {
    byte[] request =
    {
      CommandType.DIRECT_COMMAND_REPLY, NxtCommand.GET_INPUT_VALUES, (byte)portId
    };
    InputValues inputValues = new InputValues();
    byte[] reply = requestData(request);
    if (reply != null)
    {
      inputValues.inputPort = reply[3];
      // 0 is false, 1 is true.
      inputValues.valid = (reply[4] != 0);
      // 0 is false, 1 is true.
      inputValues.isCalibrated = (reply[5] == 0);
      inputValues.sensorType = reply[6];
      inputValues.sensorMode = reply[7];
      inputValues.rawADValue = (reply[8]) | ((reply[9]) << 8);
      inputValues.normalizedADValue = (reply[10]) | ((reply[11]) << 8);
      inputValues.scaledValue = (short)((reply[12]) | (reply[13] << 8));
      inputValues.calibratedValue = (short)((reply[14]) | (reply[15] << 8));
      //if (debugLevel >= DEBUG_LEVEL_MEDIUM)
      //{
      //  DebugConsole.show("DEBUG: getInputValues() returned:");
      //  inputValues.printValues();
      //}
    }

    return inputValues;
  }



    public void sendMessage(byte [] msg) throws IOException {
		if(out !=null) {
			out.write(msg.length & 0xff);
			out.write((msg.length >> 8) & 0xff);
			out.write(msg);
			out.flush();
		}
	}
	
	
	
	
}

File: android/app/src/main/java/shanksi/legonxtapplication/nxt/NxtBrick.kt
package shanksi.legonxtapplication.nxt

import android.util.Log
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

class NxtBrick(private val outputStream: OutputStream, private val inputStream: InputStream) {
    private val TAG = "NxtBrick"

    /**
     * Sends a direct command to the brick.
     * @param command the data to send
     */
    @Synchronized
    fun sendData(command: ByteArray) {
        Log.v(TAG, "sending data")
/*        if (uk.co.shanksi.nxt.LegoRobot.myClosingMode == uk.co.shanksi.nxt.LegoRobot.ClosingMode.ReleaseOnClose && uk.co.shanksi.nxt.LegoRobot.isReleased) throw RuntimeException(
            "Java frame disposed"
        )*/
        val lenMSB = byteArrayOf((command.size shr 8).toByte())
        val lenLSB = byteArrayOf(command.size.toByte())
        try {
            outputStream.write(lenLSB)
            outputStream.write(lenMSB)
            outputStream.write(command)
        } catch (ex: IOException) {
            Log.e(TAG, ex.message)
        }
    }

    /**
     * Reads the data returned as a reply from the brick.
     * @return the brick's reply
     */
    @Synchronized
    fun readData(): ByteArray? {
        Log.v(TAG, "reading data")
        var reply: ByteArray? = null
        var length = -1
        val lenMSB: Int
        var lenLSB: Int
        try {
            do lenLSB = inputStream.read() while (lenLSB < 0)
            lenMSB = inputStream.read() // MSB of reply length
            length = 0xFF and lenLSB or (0xFF and lenMSB shl 8)
            reply = ByteArray(length)
            // Rest of packet
            inputStream.read(reply)
        } catch (ex: IOException) {
            Log.e(TAG, ex.message)
        }
        return reply
    }

    /**
     * Combines sendData() and readData() in synchronized block.
     * @param request the data to send
     * @return the brick's reply
     */
    @Synchronized
    fun requestData(request: ByteArray?): ByteArray? {
        sendData(request!!)
        return readData()
    }

    /**
     * Sets the output state of a specific output port.
     */
    @Synchronized
    fun setOutputState(
        portId: Int,
        power: Byte,
        mode: Int,
        regulationMode: Byte,
        turnRatio: Int,
        runState: Byte,
        tachoLimit: Long
    ) {

        val request = byteArrayOf(
            NxtCommandType.DIRECT_COMMAND_NOREPLY,
            NxtCommand.SET_OUTPUT_STATE,
            portId.toByte(),
            power,
            mode.toByte(),
            regulationMode,
            turnRatio.toByte(),
            runState,
            tachoLimit.toByte(),
            (tachoLimit ushr 8).toByte(),
            (tachoLimit ushr 16).toByte(),
            (tachoLimit ushr 24).toByte()
        )
        sendData(request)
    }

    @Synchronized
    fun setInputMode(portId: Int, sensorType: Byte, sensorMode: Int) {
        Log.v(TAG, "setInputMode")
        val request = byteArrayOf(
            NxtCommandType.DIRECT_COMMAND_NOREPLY,
            NxtCommand.SET_INPUT_MODE,
            portId.toByte(),
            sensorType,
            sensorMode.toByte()
        )
        try {
            sendData(request)
        } catch (ex: IOException) {
            Log.e(TAG, ex.message)
        }
    }

    /**
     * Reads the values from given a sensor port.
     * @param portId the Id of the sensor port (0,..3)
     */
    @Synchronized
    fun getInputValues(portId: Int): InputValues? {
        val request = byteArrayOf(
            NxtCommandType.DIRECT_COMMAND_REPLY,
            NxtCommand.GET_INPUT_VALUES,
            portId.toByte()
        )

        val inputValues = InputValues()

        val reply = requestData(request)
        if (reply != null) {
            inputValues.inputPort = reply[3].toInt()
            // 0 is false, 1 is true.
            inputValues.valid = reply[4].toInt() != 0
            // 0 is false, 1 is true.
            inputValues.isCalibrated = reply[5].toInt() == 0
            inputValues.sensorType = reply[6].toInt()
            inputValues.sensorMode = reply[7].toInt()
            inputValues.rawADValue = reply[8].toInt() or (reply[9].toInt() shl 8)
            inputValues.normalizedADValue = reply[10].toInt() or (reply[11].toInt() shl 8)
            inputValues.scaledValue = (reply[12].toInt() or (reply[13].toInt() shl 8)).toShort()
            inputValues.calibratedValue = (reply[14].toInt() or (reply[15].toInt() shl 8)).toShort()

            inputValues.printValues(TAG)
        }
        return inputValues
    }
}